### PR TITLE
Introducing root-filtering for LogList on schema-2

### PR DESCRIPTION
**Pull request by Mercurrent**
Now with LogList2 flavour.

### CODE REVIEW

**Comment by Mercurrent** [comment]
Possible updates for LogList2 schema are in progress by @RJPercival
Happy to catch them up whenever available.

**Comment by daviddrysdale** [comment]
Please don't force-push after you've assigned a reviewer -- it loses any comments that are in progress.

**Comment by Mercurrent** [comment]
Got git(hub)-intro shot from Ivan. Hope that helps me avoiding -f.

File: loglist2/logfilter.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// the collection are treated as always compatible and included, even if
+// an empty cert chain is passed in.
+// Cert-chain when provided is expected to be full: ending with CA-cert.
+func (ll *LogList) RootCompatible(rootedChain []*x509.Certificate, roots LogRoots) LogList {
+	var compatible LogList
+	// Keep all the operators.
+	compatible.Operators = make(map[string]*Operator)
+	for opName, op := range ll.Operators {
+		compatible.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: make(map[string]*Log)}
+		copy(compatible.Operators[opName].Email, op.Email)
```
Comment (daviddrysdale) [inline loglist2/logfilter.go]: As before, you're copying some things but sharing others.

File: loglist2/logfilter.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+				compatible.Operators[opName].Logs[logName] = l
+				continue
+			}
+
+			if chainIsEmpty {
+				continue
+			}
+
+			// Check root is accepted.
+			for _, r := range roots[l.URL] {
```
Comment (daviddrysdale) [inline loglist2/logfilter.go]: Is it OK that this is linear in the number of roots for a log (~550 for a production log)?   (If not, could shift `roots` to something like `x509.CertPool` or `ctfe.PEMCertPool` that has an attached map.)

File: loglist2/logfilter.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
```
Comment (daviddrysdale) [inline loglist2/logfilter.go]: nit: this is (c) 2018 (it's mostly a copy of `../loglist/logfilter.go`, which was created in 2018)

File: loglist2/logfilter_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
```
Comment (daviddrysdale) [inline loglist2/logfilter_test.go]: (c) 2018

File: loglist2/logfilter_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+			roots: artificialRoots(testdata.TestPreCertPEM),
+			want:  subLogList(map[string]bool{"https://ct.googleapis.com/icarus/": true}), // icarus has no root info.
+		},
+		{
+			name:  "UnRootedChain",
+			in:    sampleLogList,
+			chain: singleCert(),
+			roots: artificialRoots(testdata.CACertPEM),
+			want:  subLogList(map[string]bool{}),
+		},
```
Comment (daviddrysdale) [inline loglist2/logfilter_test.go]: Missing the test for empty chain behaviour

File: loglist2/logfilter.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+			if l.State.String() != "Qualified" {
+				continue
+			}
+			active.Operators[opName].Logs[logName] = l
+		}
+	}
+	return active
+}
+
+// Compatible creates a new LogList containing only the logs of original
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply loglist2/logfilter.go]: Also, the exported method names are different than for `../loglist/logfilter.go` -- would it be easier for users if they were consistent?

File: loglist2/logfilter.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	active.Operators = make(map[string]*Operator)
+	// Keep all the operators but filter Logs.
+	for opName, op := range ll.Operators {
+		active.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: make(map[string]*Log)}
+		copy(active.Operators[opName].Email, op.Email)
+		for logName, l := range op.Logs {
+			if l.State.String() != "Qualified" {
+				continue
+			}
+			active.Operators[opName].Logs[logName] = l
```
[the remark this answers is not in the document]
Comment (Mercurrent) [reply loglist2/logfilter.go]: Moved to sharing everywhere.(hopefully)

File: loglist2/logfilter.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+// the original.
+func (ll *LogList) SelectQualified() LogList {
+	var active LogList
+	active.Operators = make(map[string]*Operator)
+	// Keep all the operators but filter Logs.
+	for opName, op := range ll.Operators {
+		active.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: make(map[string]*Log)}
+		copy(active.Operators[opName].Email, op.Email)
+		for logName, l := range op.Logs {
+			if l.State.String() != "Qualified" {
```
[the remark this answers is not in the document]
Comment (Mercurrent) [reply loglist2/logfilter.go]: Switched to Usable

File: loglist2/logfilter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mercurrent) [reply loglist2/logfilter.go]: It's an interesting question: the previous schema didn't give access to log-statuses apart from Frozen. So anything root-compatible and non-frozen would be considered as Compatible to a given Log.

With new schema we can describe more detailed Logic. And methods are less generalistic. Therefore the name.

File: loglist2/logfilter.go (hunk 1 above)
Comment (Mercurrent) [reply loglist2/logfilter.go]: Done.

File: loglist2/logfilter.go (hunk 2 above)
Comment (Mercurrent) [reply loglist2/logfilter.go]: Switched to pool. Thanks

File: loglist2/logfilter.go (hunk 3 above)
Comment (Mercurrent) [reply loglist2/logfilter.go]: Done

File: loglist2/logfilter_test.go (hunk 5 above)
Comment (Mercurrent) [reply loglist2/logfilter_test.go]: Added test.

File: loglist2/logfilter_test.go (hunk 4 above)
Comment (Mercurrent) [reply loglist2/logfilter_test.go]: Done

**Comment by Mercurrent** [comment]
PTAL

File: loglist2/logfilter.go [hunk 9]
```diff
@@ -16,24 +16,25 @@ package loglist2
...
-func (ll *LogList) SelectQualified() LogList {
+func (ll *LogList) SelectUsable() LogList {
 	var active LogList
 	active.Operators = make(map[string]*Operator)
 	// Keep all the operators but filter Logs.
 	for opName, op := range ll.Operators {
-		active.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: make(map[string]*Log)}
-		copy(active.Operators[opName].Email, op.Email)
+		active.Operators[opName] = &Operator{Logs: make(map[string]*Log)}
+		active.Operators[opName].Email = op.Email
```
Comment (daviddrysdale) [inline loglist2/logfilter.go]: nit: nicer to fold these 2 lines together
```golang
	active.Operators[opName] = &Operator{Email: op.Email, Logs: make(map[string]*Log)}
```

File: loglist2/logfilter.go [hunk 10]
```diff
@@ -53,8 +54,8 @@ func (ll *LogList) RootCompatible(rootedChain []*x509.Certificate, roots LogRoot
 	// Keep all the operators.
 	compatible.Operators = make(map[string]*Operator)
 	for opName, op := range ll.Operators {
-		compatible.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: make(map[string]*Log)}
-		copy(compatible.Operators[opName].Email, op.Email)
+		compatible.Operators[opName] = &Operator{Logs: make(map[string]*Log)}
+		compatible.Operators[opName].Email = op.Email
```
Comment (daviddrysdale) [inline loglist2/logfilter.go]: nit: fold together.

File: loglist2/logfilter_test.go [hunk 11]
```diff
@@ -27,8 +28,8 @@ func subLogList(logURLs map[string]bool) LogList {
 	var ll LogList
 	ll.Operators = make(map[string]*Operator)
 	for opName, op := range sampleLogList.Operators {
-		ll.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: map[string]*Log{}}
-		copy(ll.Operators[opName].Email, op.Email)
+		ll.Operators[opName] = &Operator{Logs: map[string]*Log{}}
+		ll.Operators[opName].Email = op.Email
```
Comment (daviddrysdale) [inline loglist2/logfilter_test.go]: nit: fold together

File: loglist2/logfilter.go (hunk 9 above)
Comment (Mercurrent) [reply loglist2/logfilter.go]: Done

File: loglist2/logfilter.go (hunk 10 above)
Comment (Mercurrent) [reply loglist2/logfilter.go]: Done

File: loglist2/logfilter_test.go (hunk 11 above)
Comment (Mercurrent) [reply loglist2/logfilter_test.go]: Done
